packages/application/src/shell.ts

```diff
@@ -1547,4 +1545,120 @@ namespace Private {
     private _stackedPanel: StackedPanel;
     private _lastCurrent: Widget | null;
   }
+
+  export class TitleWidgetHandler {
+    /**
+     * Construct a new title widget handler.
+     */
+    constructor(shell: ILabShell) {
+      this._shell = shell;
+      const titleWidget = (this._titleWidget = new Widget());
+      titleWidget.id = 'jp-title-panel-title';
+      const titleInput = document.createElement('input');
+      titleInput.type = 'text';
+      titleInput.addEventListener('keyup', this);
+      titleInput.addEventListener('click', this);
+      titleInput.addEventListener('blur', this);
+
+      titleWidget.node.appendChild(titleInput);
+    }
+
+    handleEvent(event: Event): void {
+      switch (event.type) {
+        case 'keyup':
+          void this._evtKeyUp(event as KeyboardEvent);
+          break;
+        case 'click':
+          this._evtClick(event as MouseEvent);
+          break;
+        case 'blur':
+          this._selected = false;
+          break;
+      }
+    }
+
+    /**
+     * Handle `keyup` events on the handler.
+     */
+    private async _evtKeyUp(event: KeyboardEvent): Promise<void> {
+      if (event.key == 'Enter') {
+        const widget = this._shell.currentWidget;
+        if (widget instanceof DocumentWidget) {
+          const inputElement = this.titleWidget.node.children[0] as HTMLInputElement;
+          const name = inputElement.value;
```

Comment:
This is pretty cool @cameron-toy !

Question, if the name remains the same are we still triggering a rename event? Maybe we could only rename if it is different? (on windows letter case would not really matter)

Reply:
I think not renaming if the name is the same is best. I also think some name checking needs to be done in general since I just tested and there are issues with invalid file names being entered.

Reply:
Indeed!

Reply:
Okay, filenames are now checked for validity. I also disabled selecting the title widget for non `DocumentWidget` widgets.

Reply:
Great! How does this look for the dark theme :) ? Could you post an image?

Reply:
<img width="1440" alt="Screen Shot 2021-04-22 at 9 23 54 PM" src="https://user-images.githubusercontent.com/38936057/115818174-14e4a600-a3b1-11eb-8eb2-c71c8a9938f3.png">